subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/PatternSetSnapshottingFilter.java

```diff
@@ -231,7 +231,13 @@ public int getMode() {
 
         @Override
         public FilePermissions getPermissions() {
-            int unixNumeric = stat.getUnixMode(path.toFile());
+            Path target;
+            try {
+                target = path.toRealPath();
+            } catch (IOException e) {
+                target = path;
+            }
+            int unixNumeric = stat.getUnixMode(target.toFile());
```

Comment:
:x: This code is now used in three places with slight differences. I think it would make sense to declare it once using `File` and move it to a utility method somewhere. I'd suggest moving it to the `:files` project.

Reply:
I'd say there are a lot more problems with our FileTreeElement descendants. There are copy-pasted code snippets all over the place. I don't think resolving this should be a part of this PR. This code is common for filesystem-only fileTreeElements which does not have a common ancestor. And I don't see value in adding a utility class just for that in this PR.